### DISCUSSION TITLE
ParseTree

### DISCUSSION THREAD

**Discussion by welford** [TiddlyWiki/TiddlyWiki5#5851] [Q&A] [answered]
Hello,

I'm very much struggling to navigate the parsetree and would like some help with the following problem i have run into.

I have widget for rendering [ruby elements ](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/ruby) it looks like this in a tiddler

<$r u="above">below</$r>
![image](https://user-images.githubusercontent.com/4590399/124390261-9d04f700-dce2-11eb-84df-f29d99d025c0.png)

which is generally all well and good for normal use (see: [example](http://welford.github.io/twsr.html#Answers%20without%20Scheduling)), but i am hitting problems with my implementation of this with lists:

for all tiddlers tagged fruit, make a ruby elements with tiddler name:

```
<$list filter="[tag[fruit]]">
<$r u=<<currentTiddler>>><<currentTiddler>></$r>
</$list>
```

This however renders as:

![image](https://user-images.githubusercontent.com/4590399/124390353-13095e00-dce3-11eb-9c5f-4da97a2aa249.png)

Each subsequent item picks up previous elements rubys. I have uploaded an example here: http://welford.github.io/ruby.html

The code to render this is:

```
//Render this widget into the DOM
TWSRRuby.prototype.render = function(parent,nextSibling) {
	if(!("twsrruby" in this)){
		this.parentDomNode = parent;
		this.computeAttributes();
		var upper = this.getAttribute("u");
		var useAnswer = !this.hasAttribute("na");

		var ct = this.getVariable("currentTiddler");
		var tid = $tw.wiki.getTiddler(ct);
		var tags = tid ? tid.getFieldList("tags") : [];
		if(tags.indexOf("$:/tags/twsr/noRubyAnswer") > 0){
				useAnswer = false;
		}

		//var container = this.document.createElement("div");
		//Widget.prototype.render.call(this, container, null);
		//var wikiParser = $tw.wiki.parseText("text/vnd.tiddlywiki", "<ruby>"+container.innerHTML+"<rt>"+(useAnswer?"<$a s>":"")+upper+(useAnswer?"</$a>":"")+"</rt></ruby>", {parseAsInline: true});
		var wikiParser = $tw.wiki.parseText("text/vnd.tiddlywiki", "<ruby>twsr_replace_me<rt>"+(useAnswer?"<$a s>":"")+upper+(useAnswer?"</$a>":"")+"</rt></ruby>", {parseAsInline: true});
		var newTree = wikiParser.tree;
		var temp = newTree[0].children[1];  //copy rt element out
		newTree[0].children = this.parseTreeNode.children; //replace "twsr_replace_me" with original children in <$ruby>...<\$ruby>
		newTree[0].children.push(temp); // re append rt to 
		this.parseTreeNode.children = newTree; //update children

		this.twsrruby = true;
	}
	Widget.prototype.render.call(this, parent, nextSibling);
};

```

I'm sure this is something to do with my misuse of the parseTreeNode of the containing tiddler - does anyone have pointers on how to achieve the correct behaviour here.

Thanks,
James

**Comment by welford** [comment]
In case it makes things clearer, i am trying to take this:

`<$r u="above">below</$r>`

then create the wiki parser object

`var wikiParser = $tw.wiki.parseText("text/vnd.tiddlywiki", "<ruby>twsr_replace_me<rt>"+(useAnswer?"<$a s>":"")+upper+(useAnswer?"</$a>":"")+"</rt></ruby>", {parseAsInline: true});`

which should look like this:

`"<ruby>twsr_replace_me<rt>above</rt></ruby>"`

then replace the **twsr_replace_me**  section with whatever was between <$r> and </$r> e.g. 

`"<ruby>below<rt>above</rt></ruby>"`

James

**Reply by saqimtiaz** [reply]
> "<ruby>below<rt>above</rt></ruby>"

I would set this up as a template that uses variables and the `<$list/>` widget for conditional behaviour. Then in the widget you posted above, don't bother with the parseTree and just set up the variables and transclude your template with `$tw.makeTranscludeWidget()`

**Reply by welford** [reply]
I don't follow what you mean Saq (sorry i am not very knowlegable on the insides of TW )

If i can keep the format: <$r u="above">below</$r> and change how it works in the widget to something simpler i am very willing :)

One requirement is that the sub text is able to have widgets applied to it. e.g. in this case the above section has the <$answer> widget set on it. but it could just as well be the below one.

http://welford.github.io/twsr.html#Answers%20without%20Scheduling

**Reply by welford** [reply]
I also feel like i am quite close to the correct answer here:

commenting out these lines:
```
// var temp = newTree[0].children[1];
// newTree[0].children = this.parseTreeNode.children;
// newTree[0].children.push(temp);
```

and the output is:

![image](https://user-images.githubusercontent.com/4590399/124398421-8d4ed800-dd0d-11eb-894c-1c9c0b92e227.png)

so it seems like parse tree stuff is just wrong.

**Comment by welford** [comment, marked as answer]
Went with this:

```
TWSRRuby.prototype.render = function(parent,nextSibling) {
	var oc = this.parseTreeNode.children;
	//if(!("twsrruby" in this)){
	if(true){
		this.parentDomNode = parent;
		this.computeAttributes();
		var upper = this.getAttribute("u");
		var useAnswer = !this.hasAttribute("na");

		var ct = this.getVariable("currentTiddler");
		var tid = $tw.wiki.getTiddler(ct);
		var tags = tid ? tid.getFieldList("tags") : [];
		if(tags.indexOf("$:/tags/twsr/noRubyAnswer") > 0){
				useAnswer = false;
		}

		var container = this.document.createElement("span");
		Widget.prototype.render.call(this, container, nextSibling);
		var wikiParser = $tw.wiki.parseText("text/vnd.tiddlywiki", "<ruby>"+container.innerHTML+"<rt>"+(useAnswer?"<$a s>":"")+upper+(useAnswer?"</$a>":"")+"</rt></ruby>", {parseAsInline: true});
		this.parseTreeNode.children = wikiParser.tree;
	}
	Widget.prototype.render.call(this, parent, nextSibling);
	this.parseTreeNode.children = oc;
};

```

1. store orginal children and revert
2. render widget into temp container, and insert that as a string into the wiki parser text string.
